packages/react-data/src/fetch/ReactDataFetch.js

```diff
@@ -44,50 +46,88 @@ function readResultMap(): Map<string, Result> {
   return map;
 }
 
-// TODO: options, auth, etc.
-export function fetch(url: string): Object {
-  const map = readResultMap();
-  const entry = map.get(url);
-  if (entry === undefined) {
-    let resolve = () => {};
-    const wakeable: Wakeable = new Promise(r => {
-      // TODO: should this be a plain thenable instead?
-      resolve = r;
-    });
-    const result: Result = {
-      status: Pending,
-      value: wakeable,
-    };
-    map.set(url, result);
-    const xhr = new XMLHttpRequest();
-    xhr.onload = function() {
-      // TODO: should we handle status codes?
-      if (result.status !== Pending) {
-        return;
+function createFromThenable(thenable, wrapValue): Result {
+  const result: Result = {
+    status: Pending,
+    value: (null: any),
+  };
+  result.value = thenable.then(
+    value => {
+      if (result.status === Pending) {
+        const resolvedResult = ((result: any): ResolvedResult);
+        resolvedResult.status = Resolved;
+        resolvedResult.value = wrapValue(value);
       }
-      const resolvedResult = ((result: any): ResolvedResult);
-      resolvedResult.status = Resolved;
-      resolvedResult.value = xhr.response;
-      resolve();
-    };
-    xhr.onerror = function() {
-      if (result.status !== Pending) {
-        return;
+    },
+    err => {
+      if (result.status === Pending) {
+        const rejectedResult = ((result: any): RejectedResult);
+        rejectedResult.status = Rejected;
+        rejectedResult.value = err;
       }
-      const rejectedResult = ((result: any): RejectedResult);
-      rejectedResult.status = Rejected;
-      // TODO: use something else as the error value?
-      rejectedResult.value = xhr;
-      resolve();
-    };
-    xhr.open('GET', url);
-    xhr.send();
-    throw wakeable;
-  }
-  const result: Result = entry;
+    },
+  );
+  return result;
+}
+
+function readResult(result: Result) {
   if (result.status === Resolved) {
     return result.value;
   } else {
     throw result.value;
   }
 }
+
+function Response(nativeResponse) {
+  this.headers = nativeResponse.headers;
+  this.ok = nativeResponse.ok;
+  this.redirected = nativeResponse.redirected;
+  this.status = nativeResponse.status;
+  this.statusText = nativeResponse.statusText;
+  this.type = nativeResponse.type;
+  this.url = nativeResponse.url;
+
+  this.json = function() {
```

Comment:
Let's put these on the prototype like the original. (Perhaps even two chains.)

We typically don't but we're emulating someone else's API here and the more we add, the slower it'll get to not use a prototype.

Reply:
It's probably fine to use classes that we transpile. We have a few other non-core modules that does that. Once we compile to ESM we can probably go native classes.